cpp/src/io/utilities/config_utils.cpp

```diff
@@ -80,4 +87,36 @@ bool is_stable_enabled() { return is_all_enabled() or get_env_policy() == usage_
 
 }  // namespace nvcomp_integration
 
-}  // namespace cudf::io::detail
+inline std::mutex& host_mr_lock()
+{
+  static std::mutex map_lock;
+  return map_lock;
+}
+
+inline rmm::host_async_resource_ref default_pinned_mr()
+{
+  static rmm::mr::pinned_host_memory_resource default_mr{};
+  return default_mr;
+}
+
+inline auto& host_mr()
+{
+  static rmm::host_async_resource_ref host_mr = default_pinned_mr();
+  return host_mr;
+}
+
+}  // namespace detail
+
+void set_current_host_memory_resource(rmm::host_async_resource_ref mr)
+{
+  std::lock_guard<std::mutex> lock{detail::host_mr_lock()};
+  detail::host_mr() = mr;
+}
+
+rmm::host_async_resource_ref get_current_host_memory_resource()
+{
+  std::lock_guard<std::mutex> lock{detail::host_mr_lock()};
```

Comment:
I believe we are in C++17 only, so you can drop the explicit template arguments in favor of CTAD:
```suggestion
  std::lock_guard lock{detail::host_mr_lock()};
```

Applies throughout